### ISSUE TITLE
Preceding integer assignment inconsistency

### ISSUE THREAD

**Issue by McDutchie** [ksh93/ksh#465]
Reproducer:
```
$ typeset -i NUM=123
$ (NUM=3+4 env; :)|grep ^NUM=
NUM=3+4
$ (NUM=3+4 env)|grep ^NUM=
NUM=7
```
Which output is correct, `3+4` or `7`? In other words, should the -i attribute be honoured for a preceding assignment for exporting to the env command? Bash, mksh, and zsh are all inconsistent with each other. But we need to decide on one because both of the above reproducers need to output the same value.

**Comment by hyenias** [comment]
Does  the variable NUM exist with the integer attribute applied before the `env` command  is executed (`typeset -p NUM`; NUM=3+4 env)? If it does, I believe NUM should be set to 7 during the execution of the `env` command.
```
$ zsh -c 'typeset -i NUM=123; ( typeset -p NUM; NUM=3+4 env; ) | grep NUM='
typeset -i NUM=123
NUM=7
$ zsh -c 'typeset -i NUM=123; ( typeset -p NUM; NUM=3+4 env; :) | grep NUM='
typeset -i NUM=123
NUM=7
$
$ mksh -c 'typeset -i NUM=123; ( typeset -p NUM; NUM=3+4 env; ) | grep NUM='
typeset -i NUM=123
NUM=7
$ mksh -c 'typeset -i NUM=123; ( typeset -p NUM; NUM=3+4 env; :) | grep NUM='
typeset -i NUM=123
NUM=7
$
$ ksh -c 'typeset -i NUM=123; ( typeset -p NUM; NUM=3+4 env; ) | grep NUM='
typeset -i NUM=123
NUM=7
A__z="*NUM="*SHLVL
$ ksh -c 'typeset -i NUM=123; ( typeset -p NUM; NUM=3+4 env; :) | grep NUM='
typeset -i NUM=123
NUM=3+4
```

**Comment by JohnoKing** [comment]
The bug shown in https://github.com/ksh93/ksh/issues/465#issue-1142307650 likely resides in `sh_ntfork()` or a related function. I cannot reproduce the bug when ksh is compiled with `SHOPT_SPAWN` disabled.

**Comment by hyenias** [comment]
Question: Is `NUM=3+4 env` the same thing as `export NUM=3+4; env` or is it equivalent to `NUM=3+4; export NUM; env`, or `export NUM; NUM=3+4; env`? Perhaps something different... I do not know.

**Comment by McDutchie** [comment]
It's not the same as any of those. With `NUM=3+4 env`, the exported assignment only survives while the `env` command is run. In your other three examples (which are all equivalent) the variable is permanently exported.

**Comment by McDutchie** [comment]
> The bug shown in [#465 (comment)](https://github.com/ksh93/ksh/issues/465#issue-1142307650) likely resides in `sh_ntfork()` or a related function. I cannot reproduce the bug when ksh is compiled with `SHOPT_SPAWN` disabled.

Good catch. The problem is here in `sh_ntfork()`, where the invocation-local environment list is set using a `sh_scope()` call: https://github.com/ksh93/ksh/blob/0513e760f3541bcab9819b3b9731abb839fc045d/src/cmd/ksh93/sh/xec.c#L3491-L3495

The sh_scope function is here: https://github.com/ksh93/ksh/blob/0513e760f3541bcab9819b3b9731abb839fc045d/src/cmd/ksh93/sh/name.c#L2306-L2334

Analysis:

When ksh is compiled with `SHOPT_SPAWN` disabled, external commands are always run using the regular forking mechanism. First the shell forks, then in the fork, the preceding assignments list (if any) are executed and exported in the main scope. Replacing global variables is not a problem as the variables are exported and the forked shell is then replaced by the external command using execve(2).

But when using `SHOPT_SPAWN`/`sh_ntfork()`, this cannot be done as the fork(2) use is replaced by posix_spawn(2) which does not copy the parent process into the child, therefore it's not possible to execute anything in the child before calling execve(2). Which means the preceding assignments list must be processed in the parent, not the child. Which makes overwriting global variables a no-no.

To avoid overwriting global variables, sh_ntfork() essentially repurposes the function scoping mechanism. This means the preceding assignments are treated like local variables in functions, which means they *do not inherit any attributes from the parent scope*. And that is why the integer attribute is not honoured in the buggy reproducers.

It looks like this is a design flaw in the `SHOPT_SPAWN` mechanism. It's going to be tricky to work around. Somehow the attributes from the parent scope will need to be copied into the child scope. But clearly it's possible; virtual subshells manage this, too.

**Comment by McDutchie** [comment]
As a test, removing the `NV_NOSCOPE` flag from the `nv_setlist()` invocation in `sh_scope()` fixes this bug, at the expense of introducing another one: the preceding assignment incorrectly survives the command as any pre-existing variable will be replaced in the parent scope.

```diff
diff --git a/src/cmd/ksh93/sh/name.c b/src/cmd/ksh93/sh/name.c
index d2285830e..bcf0d2386 100644
--- a/src/cmd/ksh93/sh/name.c
+++ b/src/cmd/ksh93/sh/name.c
@@ -2339,7 +2339,7 @@ void sh_scope(struct argnod *envlist, int fun)
 	{
 		dtview(newscope,(Dt_t*)sh.var_tree);
 		sh.var_tree = newscope;
-		nv_setlist(envlist,NV_EXPORT|NV_NOSCOPE|NV_IDENT|NV_ASSIGN,0);
+		nv_setlist(envlist,NV_EXPORT|NV_IDENT|NV_ASSIGN,0);
 		if(!fun)
 			return;
 		sh.var_tree = dtview(newscope,0);
```

Before:

```
$ arch/*/bin/ksh -c 'typeset -i NUM=123; (NUM=3+4 env; echo "num survives? $NUM" >&2)|grep ^NUM='
NUM=3+4
num survives? 123
```

After:

```
$ arch/*/bin/ksh -c 'typeset -i NUM=123; (NUM=3+4 env; echo "num survives? $NUM" >&2)|grep ^NUM='
NUM=7
num survives? 7
```

What we want:

```
NUM=7
num survives? 123
```

This particular `nv_setlist()` invocation is the only ones with both the `NV_EXPORT` and `NV_NOSCOPE` flags passed. So I think `nv_setlist()` will need to be fixed to handle this combination of flags by copying any pre-existing attributes from the parent scope to the new scope.

**Comment by McDutchie** [comment]
In fact, this is not just an issue for external commands.

That sh_scope() function is also used for assignments preceding a built-in command. Which is logical, as those don't create a process at all.

https://github.com/ksh93/ksh/blob/08c5e54c0c294bfdd041f524e9720c69948f03e8/src/cmd/ksh93/sh/xec.c#L1325-L1329

Which means this bug exists for them as well, regardless of whether `SHOPT_SPAWN` is compiled in.

ksh93:
```
$ /bin/ksh -c 'typeset -i NUM; NUM=3+4 command eval '\''echo $NUM'\'
3+4
```

mksh:
```
$ mksh -c 'typeset -i NUM; NUM=3+4 command eval '\''echo $NUM'\'
7
```

zsh:
```
$ zsh --emulate ksh -c 'typeset -i NUM; NUM=3+4 command eval '\''echo $NUM'\'
7
```

**Comment by McDutchie** [comment]
This should be fixed in the same place as the fix from f6bc5c03ca0160da24f39954df80e4430815890a, which fixed a similar inconsistency with the check for the read-only attribute.

**Comment by McDutchie** [comment]
In fact, the bug fixed by f6bc5c03ca0160da24f39954df80e4430815890a is simply a specific instance of this bug. The extra check for readonly was because the readonly attribute was not copied to the temporary local scope. So that fix can be replaced by the more general fix for this bug.
